examples/weather/shared/src/weather/events.rs

```diff
@@ -0,0 +1,465 @@
+use crux_core::render::render;
+use crux_core::Command;
+use serde::{Deserialize, Serialize};
+
+use crate::config::API_KEY;
+use crate::favorites::events::FavoritesEvent;
+use crate::location::capability::{get_location, is_location_enabled, LocationResponse};
+use crate::weather::model::{CurrentResponse, WEATHER_URL};
+use crate::{Effect, Event, Model};
+use crux_http::command::Http;
+
+#[derive(Serialize)]
+pub struct CurrentQueryString {
+    pub lat: String,
+    pub lon: String,
+    pub units: &'static str,
+    pub appid: String,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+pub enum WeatherEvent {
+    Show,
+    LocationEnabled(bool),
+    LocationFetched(Option<LocationResponse>),
+
+    // Events related to fetching weather data
+    #[serde(skip)]
+    Fetch(f64, f64),
+    #[serde(skip)]
+    FetchFavorite(f64, f64),
+    #[serde(skip)]
+    FetchFavorites,
+    #[serde(skip)]
+    SetWeather(Box<crux_http::Result<crux_http::Response<CurrentResponse>>>),
+    #[serde(skip)]
+    SetFavoriteWeather(
+        Box<crux_http::Result<crux_http::Response<CurrentResponse>>>,
+        f64,
+        f64,
+    ),
+}
+
+pub fn update(event: WeatherEvent, model: &mut Model) -> Command<Effect, Event> {
+    match event {
+        WeatherEvent::Show => Command::event(Event::Favorites(Box::new(FavoritesEvent::Restore)))
+            .then(
+                is_location_enabled().then_send(|result| {
+                    Event::Home(Box::new(WeatherEvent::LocationEnabled(result)))
+                }),
+            ),
+        WeatherEvent::LocationEnabled(enabled) => {
+            model.location_enabled = enabled;
+            if enabled {
+                get_location().then_send(|result| {
+                    Event::Home(Box::new(WeatherEvent::LocationFetched(result)))
+                })
+            } else {
+                Command::done()
+            }
+        }
+        WeatherEvent::LocationFetched(location) => {
+            model.last_location.clone_from(&location);
+            if let Some(loc) = location {
+                update(WeatherEvent::Fetch(loc.lat, loc.lon), model)
+            } else {
+                Command::done()
+            }
+        }
+
+        // Internal events related to fetching weather data
+        WeatherEvent::Fetch(lat, long) => Http::get(WEATHER_URL)
+            .expect_json()
+            .query(&CurrentQueryString {
+                lat: lat.to_string(),
+                lon: long.to_string(),
+                units: "metric",
+                appid: API_KEY.clone(),
+            })
+            .expect("could not serialize query string")
+            .build()
```

Comment:
A good future enhancement might be a weather API client encapsulating this request building